buildtest/buildsystem/base.py

```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+           During compiler detection, we set class variables ``self.cc``, ``self.cxx``. ``self.fc``, ``self.cflags``,
+           ``self.cxxflags``, ``self.fflags``, ``self.cppflags``. ``self.ldflags``. Finally we generate the compile
+           command and add each instruction to ``lines`` which contains content of test. Upon completion, we return
+           a list that contains content of the test.
+        """
+
+        self.compiler_recipe = self.recipe.get("compiler")
+        sourcefile = self.compiler_recipe["source"]
+        # The 'source' key can accept an absolute path or relative path with respect to Buildspec file.
+        # resolve_path returns full path to file or None if not found. We first check absolute path if not found we
+        # check for relative path.
+        self.source = resolve_path(sourcefile) or resolve_path(
+            os.path.join(
+                os.path.dirname(self.buildspec), self.compiler_recipe["source"]
+            )
+        )
+
+        # raise error if we can't find source file to compile
+        if not self.source:
+            sys.exit(f"Failed to resolve path specified by key 'source': {sourcefile}")
+
+        # name of generated executable. Add the .exe extension at end of source file name.
+        self.executable = "%s.exe" % os.path.basename(self.source)
+
+        self.compiler_name = None
+        self.lang = self.detect_lang(self.source)
+
+        # Get GNU Compiler if 'gnu' defined in compiler properties.
+        if self.compiler_recipe.get("gnu"):
+            self.compiler_name = "gnu"
+            gnu = GNUCompiler()
+            # get_compilers() will return a list that returns C, C++, and Fortran GNU wrapper
+            self.cc, self.cxx, self.fc = gnu.get_compilers()
+
+        # Get Intel Compiler
+        elif self.compiler_recipe.get("intel"):
+            self.compiler_name = "intel"
+            intel = IntelCompiler()
+            self.cc, self.cxx, self.fc = intel.get_compilers()
+
+        # Get PGI Compiler
+        elif self.compiler_recipe.get("pgi"):
+            self.compiler_name = "pgi"
+            pgi = PGICompiler()
+            self.cc, self.cxx, self.fc = pgi.get_compilers()
+
+        # Get Cray Compiler
+        elif self.compiler_recipe.get("cray"):
+            self.compiler_name = "cray"
+            cray = CrayCompiler()
+            self.cc, self.cxx, self.fc = cray.get_compilers()
+
+        # Set ldflags, cppflags, cflags, cxxflags, fflags if defined in Buildspec
+        self.ldflags = self.compiler_recipe[self.compiler_name].get("ldflags")
+        self.cppflags = self.compiler_recipe[self.compiler_name].get("cppflags")
+        self.cflags = self.compiler_recipe[self.compiler_name].get("cflags")
+        self.fflags = self.compiler_recipe[self.compiler_name].get("fflags")
+        self.cxxflags = self.compiler_recipe[self.compiler_name].get("cxxflags")
+
+        cmd = self.generate_compile_cmd()
```

Comment:
The rest of this recipe is hugely long, and should be broken into functions that can be mapped to steps. I would suggest taking these steps:

 - first refactor the logic to start with a compiler class, removing a lot of the code above with if / else because a Compiler class is just abstract to provide common functions.
 - then for each, determine what steps are common and can be shared. Write those into the Compiler parent class, and just call from the child classes. Leave the specific commands that are compiler specific.

The goal is that shared functionality is represented in neatly defined steps held by the parent Compiler Class. The children then call those steps, and it can be read from top to bottom to easily understand what is going on (it's not a big dump of spaghetti code as it is now).  Then of course each child will have some custom steps / functionality that is more prominent in the class.  I'll continue my review after this change.